server/src/main/java/org/elasticsearch/indices/recovery/RecoverySnapshotFileRequest.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.indices.recovery;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot;
+import org.elasticsearch.repositories.IndexId;
+
+import java.io.IOException;
+
+public class RecoverySnapshotFileRequest extends RecoveryTransportRequest {
+    private final long recoveryId;
+    private final ShardId shardId;
+    private final String repository;
+    private final IndexId indexId;
+    private final BlobStoreIndexShardSnapshot.FileInfo fileInfo;
+
+    public RecoverySnapshotFileRequest(long recoveryId,
+                                       long requestSeqNo,
+                                       ShardId shardId,
+                                       String repository,
+                                       IndexId indexId,
+                                       BlobStoreIndexShardSnapshot.FileInfo fileInfo) {
+        super(requestSeqNo);
+        this.recoveryId = recoveryId;
+        this.shardId = shardId;
+        this.repository = repository;
+        this.indexId = indexId;
+        this.fileInfo = fileInfo;
+    }
+
+    public RecoverySnapshotFileRequest(StreamInput in) throws IOException {
+        super(in);
+        this.recoveryId = in.readLong();
+        this.shardId = new ShardId(in);
+        this.repository = in.readString();
+        this.indexId = new IndexId(in);
+        this.fileInfo = new BlobStoreIndexShardSnapshot.FileInfo(in);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeLong(recoveryId);
```

Comment:
Can we assert that the version is new enough here? Allocation deciders should ensure that, but would be somewhat nice to assert it here.

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/76237/commits/09fa0add0295a432fea6816ceddcd58070036e9a